### PR TITLE
RavenDB-16924 various raft snapshot fixes

### PR DESCRIPTION
**Pull request by karmeli87**
### Issue link

https://issues.hibernatingrhinos.com/issue/RavenDB-16924

### Additional description

from the additional debug info we got the following:
```
Last Notified Index : 7
Last Commit Index : 6
```
Which shouldn't happened because we should notify only on indexes that were committed.

Digging into that revealed several issues that are addressed here:

- When receiving a snapshot we notified after the snapshot was committed, but before the databases reacted to the new state.
- We truncated the raft log based on the appended entries instead of the committed ones, which might result in sending unnecessary full snapshots. 
- The negotiation itself had a few edge cases which lead to sending full snapshots

### Type of change

- Bug fix

### How risky is the change?

- Moderate - High

### Backward compatibility

- Interversion test was added.

### Is it platform specific issue?

- No

### Documentation update

- No documentation update is needed 

### Testing 

- Tests have been added that prove the fix is effective or that the feature works

### Is there any existing behavior change of other features due to this change?

- No

### UI work

- No UI work is needed

### CODE REVIEW

File: src/Raven.Server/Rachis/RachisConsensus.cs [hunk 1]
```diff
@@ -300,6 +305,8 @@ private set
 
         public Action<TransactionOperationContext, CommandBase> BeforeAppendToRaftLog;
 
+        public TaskCompletionSource<Task> OnSnapshotInstalledTask;
```
Comment (ayende) [inline src/Raven.Server/Rachis/RachisConsensus.cs]: Where is this being set to a value?

Comment (karmeli87) [reply src/Raven.Server/Rachis/RachisConsensus.cs]: You are correct. I've missed that.
Simplified to use only a `Task` instead of a `TCS` and added a test
